src/main/kotlin/org/radarbase/push/integration/garmin/util/offset/OffsetRedisPersistence.kt

```diff
@@ -0,0 +1,92 @@
+package org.radarbase.push.integration.garmin.util.offset
+
+import com.fasterxml.jackson.databind.ObjectReader
+import com.fasterxml.jackson.databind.ObjectWriter
+import com.fasterxml.jackson.databind.SerializationFeature
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import org.radarbase.push.integration.garmin.util.PostponedWriter
+import org.radarbase.push.integration.garmin.util.RedisHolder
+import org.slf4j.LoggerFactory
+import java.io.IOException
+import java.nio.file.Path
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+
+/**
+ * Accesses a OffsetRange json object a Redis entry.
+ */
+class OffsetRedisPersistence(
+    private val redisHolder: RedisHolder
+) : OffsetPersistenceFactory {
+
+    override fun read(path: String): Offsets? {
+        return try {
+            redisHolder.execute { redis ->
+                redis[path]?.let { value ->
+                    redisOffsetReader.readValue<RedisOffsets>(value)
+                        .offsets
+                        .fold(Offsets(), { set, (userId, route, offset) ->
+                            set.apply { add(UserRouteOffset(userId, route, offset)) }
+                        })
+                }
+            }
+        } catch (ex: IOException) {
+            logger.error(
+                "Error reading offsets from Redis: {}. Processing all offsets.",
+                ex.toString()
+            )
+            null
+        }
+    }
+
+    override fun writer(
+        path: Path,
+        startSet: Offsets?
+    ): OffsetPersistenceFactory.Writer = RedisWriter(path, startSet)
```

Comment:
This writer is only added to once and immediately closed afterwards. So using a PostponedWriter for this is quite a heavy operation: it needs to create a thread, run the operation and close the thread again. I'd suggest not using a PostPonedWriter and instead just directly calling write. Since the read operation and write operation directly follow each other, this could be simplified to just having a `read(path: String)` and `add(path: String, offset: UserRouteOffset)` method. The `add` method can then combine reading the current offsets, adding the new offset and writing it back with the same `redisHolder`.

Reply:
Ok sure, good point